localstack/services/events/rule.py

```diff
@@ -0,0 +1,91 @@
+import re
+from typing import Optional
+
+from localstack.aws.api.events import (
+    Arn,
+    EventBusName,
+    EventPattern,
+    RoleArn,
+    RuleDescription,
+    RuleName,
+    RuleState,
+    ScheduleExpression,
+    TagList,
+)
+from localstack.services.events.models_v2 import Rule, ValidationException
+
+
+class RuleWorker:
+    def __init__(
+        self,
+        name: RuleName,
+        region: Optional[str] = None,
+        account_id: Optional[str] = None,
+        schedule_expression: Optional[ScheduleExpression] = None,
+        event_pattern: Optional[EventPattern] = None,
+        state: Optional[RuleState] = None,
+        description: Optional[RuleDescription] = None,
+        role_arn: Optional[RoleArn] = None,
+        tags: Optional[TagList] = None,
+        event_bus_name: Optional[EventBusName] = None,
+        targets: Optional[TagList] = None,
+    ):
+        self._validate_input(event_pattern, schedule_expression, event_bus_name)
+        # required to keep data and functionality separate for persistence
+        self.rule = Rule(
```

Comment:
Composition could be worth a try here. One of the downsides we should be aware of is that it introduces strong coupling between the model (changes typically coming from AWS) and our internal business logic (changes from AWS or internal reasons). Hence, whenever the model changes, we need to update the "workers" as well to match the API signature.

Reply:
the main idea is for the provider in the next iteration to not know or care about Rule as DTO and RuleService as a combination of rule-related functionality but for an abstraction on top of the AcountRegionBundle (store) to handle this. Furthermore, input validation is part of rule functionality, therefor it needs to be in RuleService, and requires to be executed before the rule is instantiated.
I don't want to extract all of these functions that don't directly act on the instance into utils or helpers because they will only be called from the RuleService and thus belong there. 
Furthermore, I don't want to pass the validation step to the provider since the provider should primarily focus on input and output handling.

Reply:
Raising exceptions or doing work in constructors is generally an anti-pattern, but I think validation exceptions can be appropriate.